span-normalizer/src/main/java/org/hypertrace/core/spannormalizer/fieldgenerators/RpcFieldsGenerator.java

```diff
@@ -0,0 +1,125 @@
+package org.hypertrace.core.spannormalizer.fieldgenerators;
+
+import io.jaegertracing.api_v2.JaegerSpanInternalModel;
+import org.apache.commons.lang3.StringUtils;
+import org.hypertrace.core.datamodel.Event;
+import org.hypertrace.core.datamodel.eventfields.grpc.Grpc;
+import org.hypertrace.core.datamodel.eventfields.rpc.Rpc;
+import org.hypertrace.core.span.constants.RawSpanConstants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+import static org.hypertrace.core.span.constants.v1.OTELRpcSystem.OTEL_RPC_SYSTEM_GRPC;
+import static org.hypertrace.core.span.constants.v1.OTELSpanTag.OTEL_SPAN_TAG_RPC_METHOD;
+import static org.hypertrace.core.span.constants.v1.OTELSpanTag.OTEL_SPAN_TAG_RPC_SERVICE;
+import static org.hypertrace.core.span.constants.v1.OTELSpanTag.OTEL_SPAN_TAG_RPC_SYSTEM;
+import static org.hypertrace.core.span.constants.v1.Rpc.RPC_REQUEST_METADATA;
+import static org.hypertrace.core.span.constants.v1.Rpc.RPC_RESPONSE_METADATA;
+
+public class RpcFieldsGenerator extends ProtocolFieldsGenerator<Rpc.Builder> {
+  private static final String OTEL_SPAN_TAG_RPC_SYSTEM_ATTR = RawSpanConstants.getValue(OTEL_SPAN_TAG_RPC_SYSTEM);
+  private static final String OTEL_RPC_SYSTEM_GRPC_VAL = RawSpanConstants.getValue(OTEL_RPC_SYSTEM_GRPC);
+  private static final char DOT = '.';
+  private static final String REQUEST_METADATA_PREFIX =
+      RawSpanConstants.getValue(RPC_REQUEST_METADATA) + DOT;
+  private static final String RESPONSE_METADATA_PREFIX =
+      RawSpanConstants.getValue(RPC_RESPONSE_METADATA) + DOT;
+  private static Logger LOGGER = LoggerFactory.getLogger(RpcFieldsGenerator.class);
+  private static Map<String, FieldGenerator<Rpc.Builder>> fieldGeneratorMap = initializeFieldGenerators();
+
+  GrpcFieldsGenerator grpcFieldsGenerator;
+
+  public RpcFieldsGenerator(GrpcFieldsGenerator grpcFieldsGenerator) {
+    this.grpcFieldsGenerator = grpcFieldsGenerator;
```

Comment:
why does this need to take an internal field generator?

Reply:
rpc.* tags will get translated to relevant grpc or other rpc protocol fields based on the value of tag `rpc.system`. For eg: rpc.request.metadata.* tags get processed by grpcFieldsGenerator after validating rpc.system = grpc.